backends/tc/backend.cpp

```diff
@@ -0,0 +1,623 @@
+                    isDefaultHit = true;
+                    defaultHit++;
+                    auto adecl = refMap->getDeclaration(action->getPath(), true);
+                    defaultActionName = externalName(adecl);
+                }
+                if (anno->name == "default_hit_const") {
+                    isDefaultHitConst = true;
+                    defaultHitConst++;
+                    auto adecl = refMap->getDeclaration(action->getPath(), true);
+                    defaultActionName = externalName(adecl);
+                }
+            }
+            if (isTableOnly && isDefaultHit && isDefaultHitConst) {
+                ::error(ErrorType::ERR_INVALID,
+                        "Table '%1%' has an action reference '%2%' which is "
+                        "annotated with '@tableonly', '@default_hit' and '@default_hit_const'",
+                        t->name.originalName, action->getName().originalName);
+                break;
+            } else if (isTableOnly && isDefaultHit) {
+                ::error(ErrorType::ERR_INVALID,
+                        "Table '%1%' has an action reference '%2%' which is "
+                        "annotated with '@tableonly' and '@default_hit'",
+                        t->name.originalName, action->getName().originalName);
+                break;
+            } else if (isTableOnly && isDefaultHitConst) {
+                ::error(ErrorType::ERR_INVALID,
+                        "Table '%1%' has an action reference '%2%' which is "
+                        "annotated with '@tableonly' and '@default_hit_const'",
+                        t->name.originalName, action->getName().originalName);
+                break;
+            } else if (isDefaultHit && isDefaultHitConst) {
+                ::error(ErrorType::ERR_INVALID,
+                        "Table '%1%' has an action reference '%2%' which is "
+                        "annotated with '@default_hit' and '@default_hit_const'",
+                        t->name.originalName, action->getName().originalName);
+                break;
+            }
+        }
+        if (::errorCount() > 0) {
+            return;
+        }
+        if ((defaultHit > 0) && (defaultHitConst > 0)) {
+            ::error(ErrorType::ERR_INVALID,
+                    "Table '%1%' cannot have both '@default_hit' action "
+                    "and '@default_hit_const' action",
+                    t->name.originalName);
+            return;
+        } else if (defaultHit > 1) {
+            ::error(ErrorType::ERR_INVALID, "Table '%1%' can have only one '@default_hit' action",
+                    t->name.originalName);
+            return;
+        } else if (defaultHitConst > 1) {
+            ::error(ErrorType::ERR_INVALID,
+                    "Table '%1%' can have only one '@default_hit_const' action",
+                    t->name.originalName);
+            return;
+        }
+        if (defaultActionName != nullptr && defaultActionName != "NoAction") {
```

Comment:
use `P4::P4CoreLibrary::instance().noAction.name`

Reply:
updated the code